source/plugin/optional/usage.rst

```diff
@@ -0,0 +1,267 @@
+======================
+
+The purpose of the ``Optional`` type is representing a value that might or might not be there. As such, many use cases
+revolve around handling absent values.
+
+Presence Check
+~~~~~~~~~~~~~~
+
+The ``isPresent()`` method returns true if a value is present on the ``Optional``. It can provide the most basic
+verification and is an equivalent to a classic ``null`` check.
+
+.. code-block:: java
+
+    Optional<String> opt = getOptionalString();
+    if (opt.isPresent()) {
+        String wrappedString = opt.get();
+        // more code
+    }
+
+Using default Values
+~~~~~~~~~~~~~~~~~~~~
+
+A common pattern is falling back to a default value if none is present. The ``orElse()`` method allows for a single
+line statement that will return either the value present on the ``Optional`` or the supplied default value.
+
+Instead of
+
+.. code-block:: java
+
+    Optional<String> optionalString = optionalString();
+    String someString;
+    if (optionalString.isPresent()) {
+        someString = optionalString.get();
+    } else {
+        someString = DEFAULT_STRING;
+    }
+
+just use
+
+.. code-block:: java
+
+    String someString = getOptionalString().orElse(DEFAULT_STRING);
+
+In some cases a default value has to be calculated in a way that has side effects or is particularly expensive. In such
+a case it is desirable to calculate the default value only if needed (*lazy evaluation*). The ``orElseGet()`` method
+accepts a ``Supplier`` instead of a pre-calculated value. If no value is present on the ``Optional`` itself, the
+``Supplier`` will be called. Since ``Supplier`` is a functional interface, a lambda expression or method reference can
+be passed instead.
+
+Instead of
+
+.. code-block:: java
+
+    Optional<String> optionalString = optionalString();
+    String someString;
+    if (optionalString.isPresent()) {
+        someString = optionalString.get();
+    } else {
+        someString = myPlugin.defaultString();
+    }
+
+just use
+
+.. code-block:: java
+
+    String someString = getOptionalString().orElseGet(myPlugin::defaultString);
+
+
+Fail on absent Values
+~~~~~~~~~~~~~~~~~~~~~
+
+If a value being absent should lead to an exception, it is almost always better to throw a custom exception instead of
+relying on the default ``NoSuchElementException``. If you call the ``orElseThrow()`` method with a ``Supplier``, it will
+return the wrapped value if it is present, or throw a ``Throwable`` obtained from the ``Supplier`` if the ``Optional``
+is empty. Again, as ``Supplier`` is a functional interface, lambda expressions or method references may be used instead.
+
+Instead of
+
+.. code-block:: java
+
+    Optional<String> optionalString = optionalString();
+    if (!optionalString.isPresent()) {
+        throw new MyException();
+    }
+    String someString = optionalString.get();
+
+just use
+
+.. code-block:: java
+
+    String someString = getOptionalString().orElseThrow();//MyException::new);
```

Comment:
Yep. My Bad. I left it this way because it breaks syntax highlighting in atom but forgot to change it back.